File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.js
function ch(p){
    if(p.length ==0) return true;
    
    let cnt = 0;
    for(let c of p){
        if(c=='(') cnt++;
        else cnt--;
        if(cnt<0) return 0;
    }
    return cnt ==0;
}
function go(p){
    if(p.length==0 || ch(p)) return p;
    
    let i=0,l=0,r=0;
    for(;i<p.length;i++){
        if(p[i]=='(')l++;
        else r++;
        if(l==r) break;
    }
    
    const a = p.substr(0,i+1);
    const b = p.substr(i+1);
    
    if(ch(a)){
        return a + go(b);
    }
    else {
        let str = "("
        str += go(b);
        str += ")";
        
        for(let i=1;i<a.length-1;i++){
            if(a[i]=='(') str += ')';
            else str += '(';
        }
        return str;
    }
}
function solution(p) {
    var answer = go(p);
    return answer;
}

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.cpp
#include <string>
#include <vector>

using namespace std;

bool ch(string s){
    int l=0;
    for(char c : s){
         if(c=='(') l++;
         else if(l) l--;
         else return false;
    }
    return (l > 0 ? false : true);
}
string convert(string s){
    string next = "(";
    for(int i=1;i<s.size()-1;i++){
         if(s[i]=='(') next += ")";
         else next += "(";
    }
    next += ")";
    return next;
}
string go(string s){
    if(ch(s)) return s;
    
    int l=0,r=0, idx =0;
    
     for(;idx<s.size();idx++){
         char c= s[idx];
        if(c=='(') l++;
        else r++;
         
        if(l==r) break;
    }
    
    string a = s.substr(0,idx+1);
    string b = s.substr(idx+1);
    
    if(ch(a)){
        return a + go(b);
    } else {
        string ret = "(";
        ret += go(b);
        ret += ")";
        for(int i=1;i<a.size()-1;i++){
             if(a[i]=='(') ret += ")";
             else ret += "(";
        }
        return ret;
    }
}
string solution(string p) {
   return go(p);
   
}